src/daemon/daemon.cpp

```diff
@@ -386,10 +395,91 @@ auto connect_rpc(mp::DaemonRpc& rpc, mp::Daemon& daemon)
+{
+    fmt::memory_buffer errors;
+    for(const auto& name : targets)
+        if(vms.find(name) == std::cend(vms))
+        {
+            if(deleted.find(name) == std::cend(deleted))
+                fmt::format_to(errors, "instance \"{}\" does not exist\n", name);
+            else
+                fmt::format_to(errors, "instance \"{}\" is deleted\n", name);
+        }
+
+    return errors.size() ? grpc_status_for(errors) : grpc::Status::OK;
+}
+
+template<typename Targets, typename InstanceMap>
+auto process_targets(const Targets& targets,
+                     const InstanceMap& vms,
+                     const InstanceMap& deleted)
+-> std::pair<std::vector<typename Targets::value_type>, grpc::Status>
+{   // TODO: use this in commands that currently duplicate the same kind of code
+    auto st = validate_targets(targets, vms, deleted);
+    auto tgts = std::vector<typename Targets::value_type>{};
+
+    if(st.ok())
+    {
+        if(targets.empty())
+            for(const auto& vm_item : vms)
+                tgts.push_back(vm_item.first);
+        else
+            std::copy(std::cbegin(targets), std::cend(targets),
+                      std::back_inserter(tgts));
+    }
+
+    return std::make_pair(tgts, st);
+}
+
+mp::SSHProcess exec_and_log(mp::SSHSession& session, const std::string& cmd)
+{
+    mpl::log(mpl::Level::debug, category, fmt::format("Executing {}.", cmd));
+    return session.exec(cmd);
+}
+
+grpc::Status stop_accepting_ssh_connections(mp::SSHSession& session)
+{
+    auto proc = exec_and_log(session, stop_ssh_cmd);
+    auto ecode = proc.exit_code();
+
+    return ecode == 0 ? grpc::Status::OK
+                      : grpc::Status{grpc::StatusCode::FAILED_PRECONDITION,
+                                     fmt::format("Could not stop sshd. '{}' exited with code {}", stop_ssh_cmd, ecode),
+                                     proc.read_std_error()};
+}
+
+grpc::Status ssh_reboot(const std::string& hostname, int port,
+                        const std::string& username, const mp::SSHKeyProvider& key_provider)
+{
+    mp::SSHSession session{hostname, port, username, key_provider};
+    stop_accepting_ssh_connections(session); /* This allows us to later detect when the machine
+    has finished restarting by waiting for SSH to be back up. Otherwise, there would be a race
+    condition, and we would be unable to distinguish whether it had ever been down. */
+
+    auto proc = exec_and_log(session, reboot_cmd);
+    try
+    {
+        auto ecode = proc.exit_code();
+        return grpc::Status{grpc::StatusCode::FAILED_PRECONDITION,
+                            fmt::format("Reboot command exited with code {}", ecode),
+                            proc.read_std_error()}; /* we shouldn't get this far: a
+                            successful reboot command does not return */
+    }
+    catch(const mp::ExitlessSSHProcessException&) { /* this is the expected path */ }
```

Comment:
C++ style comments please

Reply:
In this case I will use a reconsideration plea :)

I thought it was nice to show that the comment "belongs" to the scope of the catch, although perhaps not worth a 3 or 4-line block:
```
catch
{
    // this is the expected path
}
```
And the two-lines alternative looks strange:

```
catch {   // this is the expected path
}
```

In general, do you prefer no c-style comments at all, in any case? For instance, I also find that clarity and brevity are sometimes improved by localized comments, say, in rvalue function arguments: `f(/* use_cache = */ true); ` (a variable could be used here, but that arguably adds to the noise). Or to signal unused parameters that you still want to name: `void f(Abc& /* alphabet */)`.

Let me know what the policy is and I will react accordingly.

Reply:
Well, clang-format is going to muck with this any ways, and there other places in the code where we do something similar to:
```
catch
{
    // this is the expected path
}
```
So it's probably better to just change it to be consistent.

As far as c-style comments, I'm not opposed to them, but we should use them in a consistent manner, like for example, the article you pointed out for reasons to use them.  But if we do that, it should be used in all of the code and introduced as a separate PR.  Also, we use c-style comments for our copyright headers;)

Reply:
All right, thanks. Will do that.